apps/laboratory/src/hooks/useWalletConnectCosigner.ts

```diff
@@ -0,0 +1,238 @@
+   */
+  async function updatePermissionsContext(
+    address: string,
+    projectId: string,
+    updateData: UpdatePermissionsContextRequest
+  ): Promise<void> {
+    const url = `${baseUrl}/${encodeURIComponent(address)}/context`
+
+    try {
+      await axios.post(url, updateData, {
+        params: { projectId },
+        headers: { 'Content-Type': 'application/json' }
+      })
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        const axiosError = error as AxiosError
+        if (axiosError.response) {
+          if (axiosError.response.status === 401) {
+            throw new CoSignerApiError(401, 'Unauthorized - Wrong signature')
+          }
+          throw new CoSignerApiError(axiosError.response.status, axiosError.response.data as string)
+        } else {
+          throw new CoSignerApiError(500, 'Network error')
+        }
+      }
+      // Re-throw if it's not an Axios error
+      throw error
+    }
+  }
+
+  /**
+   * Revokes a permission from account sessions.
+   *
+   * @param address - CAIP-10 address format
+   * @param projectId - The project identifier
+   * @param revokeData - The revoke data including pci and signature
+   * @returns A promise that resolves when the revocation is successful
+   * @throws {CoSignerApiError} If the API request fails
+   */
+  async function revokePermission(
+    address: string,
+    projectId: string,
+    revokeData: RevokePermissionRequest
+  ): Promise<void> {
+    const url = `${baseUrl}/${encodeURIComponent(address)}/revoke`
+
+    try {
+      await axios.post(url, revokeData, {
+        params: { projectId },
+        headers: { 'Content-Type': 'application/json' }
+      })
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        const axiosError = error as AxiosError
+        if (axiosError.response) {
+          if (axiosError.response.status === 401) {
+            throw new CoSignerApiError(401, 'Unauthorized - Wrong signature')
+          }
+          throw new CoSignerApiError(axiosError.response.status, axiosError.response.data as string)
+        } else {
+          throw new CoSignerApiError(500, 'Network error')
+        }
+      }
+      // Re-throw if it's not an Axios error
+      throw error
+    }
+  }
+  /**
+   * Sends a co-signing request for a user operation.
+   *
+   * @param address - CAIP-10 address format
+   * @param projectId - The project identifier
+   * @param coSignData - The co-sign data including pci and userOp
+   * @returns A promise that resolves to the user operation receipt
+   * @throws {CoSignerApiError} If the API request fails
+   */
+  async function coSignUserOperation(
+    address: string,
+    projectId: string,
+    coSignData: CoSignRequest
+  ): Promise<CoSignResponse> {
+    const url = `${baseUrl}/${encodeURIComponent(address)}/sign`
+
+    try {
+      const response = await axios.post<CoSignResponse>(url, coSignData, {
+        params: { projectId },
```

Comment:
seems most code is repeated in these functions. I'd abstract to some common api util or similar and call it with different params from here

Reply:
Applies for all of the functions in this file

Reply:
I think its already abstracted, will not affect much if tried to abstract it more.

Reply:
@KannuSingh I think @tomiir was referencing the request code specifically. If you check your methods that interact with co-signer, you will notice that about 90% of the code is the same. The only difference is the request. This can be handled with a new method to interact with co-signer and usage of generics to provide type

Reply:
This should be extracted to a util, named something like `sendCoSignerRequest<RequestType>` or similar

Reply:
extracted the logic and create the similar method